allennlp/training/trainer.py

```diff
@@ -881,25 +877,27 @@ def from_partial_objects(
         if not optimizer_:
             optimizer_ = Optimizer.default(parameters)
 
-        batches_per_epoch = iterator.get_num_batches(train_data)
-        if batches_per_epoch == 1:  # get_num_batches returns 1 when it can't determine the answer
+        try:
```

Comment:
I thought you gave an implementation.  Oh, this is the data loader, not the dataset...  Ok.  But if we never call `len()` on the dataset itself, we don't need a default implementation there anymore, do we?  Or does one of the samplers call `len()` on the dataset?

Reply:
The dataloader unfortunately calls len (and warns if you call len on a lazy dataset, but still requires it to be implemented, somewhat bizarrely).